src/tf/tf_parser.cpp

```diff
@@ -286,9 +286,30 @@ void tf_parser::parse_graph(const tensorflow::GraphDef& graph)
     {
         this->parse_node(p.first);
     }
-
-    // Needs to add a ret instruction at the end of
-    // the program
+    auto last_ins = std::prev(mm->end());
+    if(last_ins != mm->end())
+    {
+        // Needs to add a ret instruction at the end of
+        // the program
+        if(output_node_names.empty())
```

Comment:
There is no way to compute output_node_name, by just taking the nodes that are never used for input?

Reply:
There could be nodes that remain from training that don't have the output being used. I think that's the main issue with checking if a node is never used for input.

Reply:
I think you can compute the outputs with something like this:

```cpp
std::unordered_set<std::string> inputs;
for(auto&& p : nodes)
{
    auto&& node = p.second;
    std::copy(node.input().begin(), node.input().end(), std::inserter(inputs, inputs.end()));
}
std::unordered_set<std::string> outputs;
std::copy_if(nodes.begin(), nodes.end(), std::inserter(outputs, outputs.end()), [&](auto&& p) {
    const auto& name = p.first;
    const auto& node = p.second;
    if(contains(name, "^"))
        return false;
    for(const auto& x:{"NoOp", "Assert"})
    {
        if (node.op() == x)
            return false;
        if(contains(name, x))
            return false;
    }
    return inputs.count(name) == 0;
});
```

Reply:
So when the user doesn't specify the outputs how to pick which output? Or is it random?

Reply:
In TF, passing in the output name is enforced when running the graph or exporting it. Otherwise, TF will throw an error.

Reply:
Then where can a user get the output names?

Reply:
They can use TF scripts that read pb files (similar to onnx), then pick the output node they want.